File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/SiteProvider.java
package com.wa2c.android.medoly.plugin.action.lyricsscraper;

import android.content.ContentProvider;
import android.content.ContentProviderOperation;
import android.content.ContentProviderResult;
import android.content.ContentValues;
import android.content.Context;
import android.content.OperationApplicationException;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.database.sqlite.SQLiteStatement;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import com.wa2c.android.medoly.utils.Logger;

import java.util.ArrayList;

public class SiteProvider extends ContentProvider {
    /** 再生キューテーブル。 */
    private static final String SITE_TABLE_NAME = "sites";
    /** プレイリストマップテーブル。 */
    private static final String GROUP_TABLE_NAME = "groups";

    /** URIのauthority。 */
    private static final String AUTHORITY = "com.wa2c.android.medoly.plugin.action.lyricsscraper.siteprovider";
    /** コンテンツアクセスURI。 */
    private static final Uri CONTENT_URI = Uri.parse("content://" + AUTHORITY);

    /** URIマッチング。 */
    private static final UriMatcher uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
    /** 再生キューURIコード。 */
    private static final int SITE_URI_CODE = 1;
    /** プレイリストURIコード。 */
    private static final int GROUP_URI_CODE = 2;

    static {
        uriMatcher.addURI(AUTHORITY, SITE_TABLE_NAME, SITE_URI_CODE);
        uriMatcher.addURI(AUTHORITY, GROUP_TABLE_NAME, GROUP_URI_CODE);
    }

    /** 再生キューアクセスURI。 */
    public static final Uri SITE_URI = Uri.withAppendedPath(CONTENT_URI, SITE_TABLE_NAME);
    /** プレイリストアクセスURI。 */
    public static final Uri GROUP_URI = Uri.withAppendedPath(CONTENT_URI, GROUP_TABLE_NAME);

    private static String getTableName(Uri uri) {
        String tableName = null;
        switch (uriMatcher.match(uri)) {
            case SITE_URI_CODE:
                tableName = SITE_TABLE_NAME;
                break;
            case GROUP_URI_CODE:
                tableName = GROUP_TABLE_NAME;
                break;
            default:
                Logger.w("Unsupported uri: '" + uri + "'.");
                break;
        }
        return tableName;
    }

    /** データベースヘルパー。 */
    private DBHelper dbHelper;

    @Override
    public boolean onCreate() {
        dbHelper = DBHelper.getInstance(getContext());
        return true;
    }

    @Override
    public Cursor query(@NonNull Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {
        String tableName = getTableName(uri);
        if (TextUtils.isEmpty(tableName))
            return null;

        SQLiteDatabase db = dbHelper.getReadableDatabase();
        try {
            return db.query(tableName, projection, selection, selectionArgs, null, null, sortOrder);
        } catch (Exception e) {
            Logger.e(e);
            return null;
        }
    }

    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }

    @Override
    public Uri insert(@NonNull Uri uri, ContentValues values) {
        String tableName = getTableName(uri);
        if (TextUtils.isEmpty(tableName))
            return null;

        SQLiteDatabase db = dbHelper.getWritableDatabase();
        try {
            long id = db.insertOrThrow(tableName, null, values);
            return Uri.withAppendedPath(uri, String.valueOf(id));
        } catch (Exception e) {
            Logger.e(e);
            return null;
        }
    }

    @Override
    public int delete(@NonNull Uri uri, String selection, String[] selectionArgs) {
        String tableName = getTableName(uri);
        if (TextUtils.isEmpty(tableName))
            return -1;

        SQLiteDatabase db = dbHelper.getWritableDatabase();
        try {
            if (selection == null && selectionArgs == null) {
                // 全削除の場合はシーケンス番号リセット
                db.execSQL("DELETE FROM SQLITE_SEQUENCE WHERE NAME = '" + tableName + "'");
            }
            return db.delete(tableName, selection, selectionArgs);
        } catch (Exception e) {
            Logger.e(e);
            return -1;
        }
    }

    @Override
    public int update(@NonNull Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        String tableName = getTableName(uri);
        if (TextUtils.isEmpty(tableName))
            return -1;

        SQLiteDatabase db = dbHelper.getWritableDatabase();
        try {
            return db.update(tableName, values, selection, selectionArgs);
        } catch (Exception e) {
            Logger.e(e);
            return -1;
        }
    }

    @NonNull
    @Override
    public ContentProviderResult[] applyBatch(@NonNull ArrayList<ContentProviderOperation> operations) throws OperationApplicationException {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        db.beginTransaction();
        try {
            ContentProviderResult[] result = super.applyBatch(operations);
            db.setTransactionSuccessful();
            return result;
        } finally {
            db.endTransaction();
        }
    }

    @Override
    public int bulkInsert(@NonNull Uri uri, @NonNull ContentValues[] values) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        db.beginTransaction();

        try {
            int insertCount = 0;
            boolean isFirst = true;
            StringBuilder insertField = new StringBuilder();
            StringBuilder insertParam = new StringBuilder();
            switch (uriMatcher.match(uri)) {
                case SITE_URI_CODE:
                    // コンパイルステートメント作成
                    SiteColumn[] siteColumns = SiteColumn.values();
                    for (SiteColumn col : siteColumns) {
                        if (isFirst) {
                            insertField.append(col.getColumnKey());
                            insertParam.append("?");
                            isFirst = false;
                        } else {
                            insertField.append(",").append(col.getColumnKey());
                            insertParam.append(",?");
                        }
                    }
                    SQLiteStatement siteStatement = db.compileStatement(
                            "INSERT INTO " + SITE_TABLE_NAME + " (" + insertField.toString() + ") VALUES (" + insertParam.toString() + ");");

                    // 挿入
                    for (ContentValues value : values) {
                        siteStatement.clearBindings();
                        int index = 1;
                        for (SiteColumn col : siteColumns) {
                            String val = value.getAsString(col.getColumnKey());
                            if (val != null) siteStatement.bindString(index, val);
                            index++;
                        }
                        if (siteStatement.executeInsert() > 0) {
                            insertCount++;
                        }
                    }

                    db.setTransactionSuccessful();
                    return insertCount;
                case GROUP_URI_CODE:
                    // コンパイルステートメント作成
                    GroupColumn[] groupColumns = GroupColumn.values();
                    for (GroupColumn col : groupColumns) {
                        if (isFirst) {
                            insertField.append(col.getColumnKey());
                            insertParam.append("?");
                            isFirst = false;
                        } else {
                            insertField.append(",").append(col.getColumnKey());
                            insertParam.append(",?");
                        }
                    }
                    SQLiteStatement groupStatement = db.compileStatement(
                            "INSERT INTO " + GROUP_TABLE_NAME + " (" + insertField.toString() + ") VALUES (" + insertParam.toString() + ");");

                    // 挿入
                    for (ContentValues value : values) {
                        groupStatement.clearBindings();
                        int index = 1;
                        for (GroupColumn col : groupColumns) {
                            String val = value.getAsString(col.getColumnKey());
                            if (val != null) groupStatement.bindString(index, val);
                            index++;
                        }
                        if (groupStatement.executeInsert() > 0) {
                            insertCount++;
                        }
                    }

                    db.setTransactionSuccessful();
                    return insertCount;
                default:
                    Logger.w("Unsupported uri: '" + uri + "'.");
                    return -1;
            }
        } catch (Exception e) {
            Logger.e(e);
            return -1;
        } finally {
            db.endTransaction();
        }
    }

    /**
     * DB Open Helper
     */
    private static class DBHelper extends SQLiteOpenHelper {

        /** データベース名。 */
        private static final String DB_NAME = "medoly_lyricsscraper.db";
        /** データベースバージョン。
         * 1: 初期、再生キューテーブル作成
         */
        private static final int DB_VERSION = 1;
        /** データベースオブジェクト。 */
        private static DBHelper helper = null;

        /**
         * インスタンスを取得。シングルトンパターンを使用する。
         * @param context コンテキスト。
         * @return インスタンス。
         */
        public static synchronized DBHelper getInstance(Context context) {
            if (helper == null) {
                helper = new DBHelper(context);
            }
            return helper;
        }

        /**
         * コンストラクタ。
         * @param context コンテキスト。
         */
        private DBHelper(Context context) {
            super(context, DB_NAME, null, DB_VERSION);
        }

        /**
         * DB作成時。
         */
        @Override
        public void onCreate(SQLiteDatabase db) {
            createSiteTable(db);
            createGroupTable(db);
        }

        /**
         * Create site table.
         * @param db  database.
         */
        private void createSiteTable(SQLiteDatabase db) {
            // Make constraint
            StringBuilder builder = new StringBuilder();
            builder.append("CREATE TABLE IF NOT EXISTS " + SITE_TABLE_NAME + " (");
            SiteColumn[] columns = SiteColumn.values();
            for (int i = 0; i < columns.length; i++) {
                if (i != 0) builder.append(",");
                builder.append(columns[i].getConstraint());
            }
            builder.append(");");

            // Create table
            db.execSQL(builder.toString());

            // Create index
            db.execSQL("CREATE INDEX IF NOT EXISTS " + SiteColumn.SITE_ID.getColumnKey() + "_idx on " + SITE_TABLE_NAME + "(" + SiteColumn.SITE_ID.getColumnKey() + ")");
            db.execSQL("CREATE INDEX IF NOT EXISTS " + SiteColumn.GROUP_ID.getColumnKey() + "_idx on " + SITE_TABLE_NAME + "(" + SiteColumn.GROUP_ID.getColumnKey() + ")");
            db.execSQL("CREATE INDEX IF NOT EXISTS " + SiteColumn.SITE_NAME.getColumnKey() + "_idx on " + SITE_TABLE_NAME + "(" + SiteColumn.SITE_NAME.getColumnKey() + ")");
        }

        /**
         * Create group table.
         * @param db database.
         */
        private void createGroupTable(SQLiteDatabase db) {
            // Make constraint
            StringBuilder builder = new StringBuilder();
            builder.append("CREATE TABLE IF NOT EXISTS " + GROUP_TABLE_NAME + " (");
            GroupColumn[] columns = GroupColumn.values();
            for (int i = 0; i < columns.length; i++) {
                if (i != 0) builder.append(",");
                builder.append(columns[i].getConstraint());
            }
            builder.append(");");

            // Create table
            db.execSQL(builder.toString());

            // Create index
            db.execSQL("CREATE INDEX IF NOT EXISTS " + GroupColumn.GROUP_ID.getColumnKey() + "_idx on " + GROUP_TABLE_NAME + "(" + GroupColumn.GROUP_ID.getColumnKey() + ")");
            db.execSQL("CREATE INDEX IF NOT EXISTS " + GroupColumn.NAME.getColumnKey() + "_idx on " + GROUP_TABLE_NAME + "(" + GroupColumn.NAME.getColumnKey() + ")");
        }

        /**
         * DBアップデート時。
         */
        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        }

    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lyricsscraper/db/SiteProvider.kt
package com.wa2c.android.medoly.plugin.action.lyricsscraper.db

import android.content.ContentProvider
import android.content.ContentProviderOperation
import android.content.ContentProviderResult
import android.content.ContentValues
import android.content.Context
import android.content.OperationApplicationException
import android.content.UriMatcher
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteStatement
import android.net.Uri
import android.text.TextUtils

import com.wa2c.android.medoly.plugin.action.lyricsscraper.util.Logger

import java.util.ArrayList

class SiteProvider : ContentProvider() {

    /** データベースヘルパー。  */
    private var dbHelper: DBHelper? = null

    override fun onCreate(): Boolean {
        dbHelper = DBHelper.getInstance(context)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        val tableName = getTableName(uri)
        if (tableName.isNullOrEmpty())
            return null

        val db = dbHelper!!.readableDatabase
        return try {
            db.query(tableName, projection, selection, selectionArgs, null, null, sortOrder)
        } catch (e: Exception) {
            Logger.e(e)
            null
        }

    }

    override fun getType(uri: Uri): String? {
        return null
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val tableName = getTableName(uri)
        if (tableName.isNullOrEmpty())
            return null

        val db = dbHelper!!.writableDatabase
        try {
            val id = db.insertOrThrow(tableName, null, values)
            return Uri.withAppendedPath(uri, id.toString())
        } catch (e: Exception) {
            Logger.e(e)
            return null
        }

    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        val tableName = getTableName(uri)
        if (tableName.isNullOrEmpty())
            return -1

        val db = dbHelper!!.writableDatabase
        return try {
            if (selection == null && selectionArgs == null) {
                // 全削除の場合はシーケンス番号リセット
                db.execSQL("DELETE FROM SQLITE_SEQUENCE WHERE NAME = '$tableName'")
            }
            db.delete(tableName, selection, selectionArgs)
        } catch (e: Exception) {
            Logger.e(e)
            -1
        }

    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        val tableName = getTableName(uri)
        if (tableName.isNullOrEmpty())
            return -1

        val db = dbHelper!!.writableDatabase
        return try {
            db.update(tableName, values, selection, selectionArgs)
        } catch (e: Exception) {
            Logger.e(e)
            -1
        }

    }

    @Throws(OperationApplicationException::class)
    override fun applyBatch(operations: ArrayList<ContentProviderOperation>): Array<ContentProviderResult?> {
        val db = dbHelper!!.writableDatabase
        db.beginTransaction()
        var result = arrayOfNulls<ContentProviderResult>(0)
        try {
            result = super.applyBatch(operations)
            db.setTransactionSuccessful()
        } catch (e: Exception) {
            Logger.e(e)
        }

        db.endTransaction()
        return result
    }

    override fun bulkInsert(uri: Uri, values: Array<ContentValues>): Int {
        val db = dbHelper!!.writableDatabase
        db.beginTransaction()

        try {
            var insertCount = 0
            var isFirst = true
            val insertField = StringBuilder()
            val insertParam = StringBuilder()
            when (uriMatcher.match(uri)) {
                SITE_URI_CODE -> {
                    // コンパイルステートメント作成
                    val siteColumns = SiteColumn.values()
                    for (col in siteColumns) {
                        if (isFirst) {
                            insertField.append(col.columnKey)
                            insertParam.append("?")
                            isFirst = false
                        } else {
                            insertField.append(",").append(col.columnKey)
                            insertParam.append(",?")
                        }
                    }
                    val siteStatement = db.compileStatement(
                            "INSERT INTO " + SITE_TABLE_NAME + " (" + insertField.toString() + ") VALUES (" + insertParam.toString() + ");")

                    // 挿入
                    for (value in values) {
                        siteStatement.clearBindings()
                        var index = 1
                        for (col in siteColumns) {
                            val `val` = value.getAsString(col.columnKey)
                            if (`val` != null) siteStatement.bindString(index, `val`)
                            index++
                        }
                        if (siteStatement.executeInsert() > 0) {
                            insertCount++
                        }
                    }

                    db.setTransactionSuccessful()
                    return insertCount
                }
                GROUP_URI_CODE -> {
                    // コンパイルステートメント作成
                    val groupColumns = GroupColumn.values()
                    for (col in groupColumns) {
                        if (isFirst) {
                            insertField.append(col.columnKey)
                            insertParam.append("?")
                            isFirst = false
                        } else {
                            insertField.append(",").append(col.columnKey)
                            insertParam.append(",?")
                        }
                    }
                    val groupStatement = db.compileStatement(
                            "INSERT INTO " + GROUP_TABLE_NAME + " (" + insertField.toString() + ") VALUES (" + insertParam.toString() + ");")

                    // 挿入
                    for (value in values) {
                        groupStatement.clearBindings()
                        var index = 1
                        for (col in groupColumns) {
                            val `val` = value.getAsString(col.columnKey)
                            if (`val` != null) groupStatement.bindString(index, `val`)
                            index++
                        }
                        if (groupStatement.executeInsert() > 0) {
                            insertCount++
                        }
                    }

                    db.setTransactionSuccessful()
                    return insertCount
                }
                else -> {
                    Logger.w("Unsupported uri: '$uri'.")
                    return -1
                }
            }
        } catch (e: Exception) {
            Logger.e(e)
            return -1
        } finally {
            db.endTransaction()
        }
    }

    /**
     * DB Open Helper
     */
    private class DBHelper private constructor(context: Context) : SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {

        /**
         * DB作成時。
         */
        override fun onCreate(db: SQLiteDatabase) {
            createSiteTable(db)
            createGroupTable(db)
        }

        /**
         * Create site table.
         * @param db  database.
         */
        private fun createSiteTable(db: SQLiteDatabase) {
            // Make constraint
            val builder = StringBuilder()
            builder.append("CREATE TABLE IF NOT EXISTS $SITE_TABLE_NAME (")
            val columns = SiteColumn.values()
            for (i in columns.indices) {
                if (i != 0) builder.append(",")
                builder.append(columns[i].getConstraint())
            }
            builder.append(");")

            // Create table
            db.execSQL(builder.toString())

            // Create index
            db.execSQL("CREATE INDEX IF NOT EXISTS " + SiteColumn.SITE_ID.columnKey + "_idx on " + SITE_TABLE_NAME + "(" + SiteColumn.SITE_ID.columnKey + ")")
            db.execSQL("CREATE INDEX IF NOT EXISTS " + SiteColumn.GROUP_ID.columnKey + "_idx on " + SITE_TABLE_NAME + "(" + SiteColumn.GROUP_ID.columnKey + ")")
            db.execSQL("CREATE INDEX IF NOT EXISTS " + SiteColumn.SITE_NAME.columnKey + "_idx on " + SITE_TABLE_NAME + "(" + SiteColumn.SITE_NAME.columnKey + ")")
        }

        /**
         * Create group table.
         * @param db database.
         */
        private fun createGroupTable(db: SQLiteDatabase) {
            // Make constraint
            val builder = StringBuilder()
            builder.append("CREATE TABLE IF NOT EXISTS $GROUP_TABLE_NAME (")
            val columns = GroupColumn.values()
            for (i in columns.indices) {
                if (i != 0) builder.append(",")
                builder.append(columns[i].getConstraint())
            }
            builder.append(");")

            // Create table
            db.execSQL(builder.toString())

            // Create index
            db.execSQL("CREATE INDEX IF NOT EXISTS " + GroupColumn.GROUP_ID.columnKey + "_idx on " + GROUP_TABLE_NAME + "(" + GroupColumn.GROUP_ID.columnKey + ")")
            db.execSQL("CREATE INDEX IF NOT EXISTS " + GroupColumn.NAME.columnKey + "_idx on " + GROUP_TABLE_NAME + "(" + GroupColumn.NAME.columnKey + ")")
        }

        /**
         * DBアップデート時。
         */
        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {}

        companion object {

            /** データベース名。  */
            private const val DB_NAME = "medoly_lyricsscraper.db"
            /** データベースバージョン。
             * 1: 初期、再生キューテーブル作成
             */
            private const val DB_VERSION = 1
            /** データベースオブジェクト。  */
            private var helper: DBHelper? = null

            /**
             * インスタンスを取得。シングルトンパターンを使用する。
             * @param context コンテキスト。
             * @return インスタンス。
             */
            @Synchronized
            fun getInstance(context: Context): DBHelper {
                if (helper == null) {
                    helper = DBHelper(context)
                }
                return helper!!
            }
        }

    }

    companion object {
        /** 再生キューテーブル。  */
        private const val SITE_TABLE_NAME = "sites"
        /** プレイリストマップテーブル。  */
        private const val GROUP_TABLE_NAME = "groups"

        /** URIのauthority。  */
        private const val AUTHORITY = "com.wa2c.android.medoly.plugin.action.lyricsscraper.siteprovider"
        /** コンテンツアクセスURI。  */
        private val CONTENT_URI = Uri.parse("content://" + AUTHORITY)

        /** URIマッチング。  */
        private val uriMatcher = UriMatcher(UriMatcher.NO_MATCH)
        /** 再生キューURIコード。  */
        private const val SITE_URI_CODE = 1
        /** プレイリストURIコード。  */
        private const val GROUP_URI_CODE = 2

        init {
            uriMatcher.addURI(AUTHORITY, SITE_TABLE_NAME, SITE_URI_CODE)
            uriMatcher.addURI(AUTHORITY, GROUP_TABLE_NAME, GROUP_URI_CODE)
        }

        /** 再生キューアクセスURI。  */
        val SITE_URI = Uri.withAppendedPath(CONTENT_URI, SITE_TABLE_NAME)!!
        /** プレイリストアクセスURI。  */
        val GROUP_URI = Uri.withAppendedPath(CONTENT_URI, GROUP_TABLE_NAME)!!

        private fun getTableName(uri: Uri): String? {
            var tableName: String? = null
            when (uriMatcher.match(uri)) {
                SITE_URI_CODE -> tableName = SITE_TABLE_NAME
                GROUP_URI_CODE -> tableName = GROUP_TABLE_NAME
                else -> Logger.w("Unsupported uri: '$uri'.")
            }
            return tableName
        }
    }

}


